File: analysis/compute_volume.R
library(jsonlite)
library(ggplot2)

data_dir <- "/Users/williamydh/Library/Mobile Documents/com~apple~CloudDocs/Sync_iCloud/Career/WT/Phoenix_Capital/Quant/kalshi_tracker/data/kalshi_trades"
output_dir <- "/Users/williamydh/Library/Mobile Documents/com~apple~CloudDocs/Sync_iCloud/Career/WT/Phoenix_Capital/Quant/kalshi_tracker/analysis"

if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
}

trade_files <- sort(list.files(data_dir, pattern = "\\.json$", full.names = TRUE))

if (length(trade_files) == 0) {
  stop("No trade files found in ", data_dir)
}

compute_daily_volume <- function(file_path) {
  payload <- fromJSON(file_path, simplifyVector = FALSE)
  date <- as.Date(payload$date)
  trades <- payload$trades
  total_volume <- sum(vapply(trades, function(trade) trade$count, numeric(1)))
  data.frame(date = date, total_volume = total_volume)
}

daily_volumes <- do.call(
  rbind,
  lapply(trade_files, compute_daily_volume)
)

daily_volumes <- daily_volumes[order(daily_volumes$date), ]
rownames(daily_volumes) <- NULL

rolling_window <- 7
rolling_total <- vapply(
  seq_len(nrow(daily_volumes)),
  function(idx) {
    start_idx <- max(1, idx - rolling_window + 1)
    sum(daily_volumes$total_volume[start_idx:idx])
  },
  numeric(1)
)

daily_volumes$rolling_total <- rolling_total

write.csv(daily_volumes, file = file.path(output_dir, "daily_total_volume.csv"), row.names = FALSE)

format_millions <- function(x) {
  paste0(format(round(x / 1e6, 1), nsmall = 1), "M")
}

plot_path <- file.path(output_dir, "7d_rolling_total_volume.png")

plt <- ggplot(daily_volumes, aes(x = date, y = rolling_total)) +
  geom_line(color = "#0b3d91", linewidth = 1.1) +
  geom_point(color = "#0b3d91", size = 1.9) +
  labs(
    title = "7-Day Rolling Total Trading Volume",
    x = "Date",
    y = "7-Day Total Volume (Contracts)"
  ) +
  scale_y_continuous(labels = format_millions) +
  theme_minimal(base_size = 14) +
  theme(
    panel.grid.major = element_line(color = "grey85"),
    panel.grid.minor = element_blank(),
    plot.title = element_text(face = "bold")
  )

ggsave(plot_path, plt, width = 14, height = 9, dpi = 150)

message("Daily totals saved to: ", file.path(output_dir, "daily_total_volume.csv"))
message("7-day rolling total chart saved to: ", plot_path)

File: analysis/compute_volume.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, List

import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
import pandas as pd


DATA_DIR = Path(
    "/Users/williamydh/Library/Mobile Documents/com~apple~CloudDocs/Sync_iCloud/Career/WT/Phoenix_Capital/Quant/kalshi_tracker/data/kalshi_trades"
)
OUTPUT_DIR = Path(
    "/Users/williamydh/Library/Mobile Documents/com~apple~CloudDocs/Sync_iCloud/Career/WT/Phoenix_Capital/Quant/kalshi_tracker/analysis"
)


@dataclass
class DailyVolume:
    date: pd.Timestamp
    total_volume: int


def read_daily_volume(file_path: Path) -> DailyVolume:
    with file_path.open("r", encoding="utf-8") as f:
        payload = json.load(f)

    date = pd.to_datetime(payload["date"])
    trades: Iterable[dict] = payload.get("trades", [])
    total_volume = int(sum(trade.get("count", 0) for trade in trades))
    return DailyVolume(date=date, total_volume=total_volume)


def collect_daily_volumes(files: Iterable[Path]) -> pd.DataFrame:
    volumes: List[DailyVolume] = [read_daily_volume(path) for path in files]
    df = pd.DataFrame([vars(item) for item in volumes])
    df.sort_values("date", inplace=True)
    df.reset_index(drop=True, inplace=True)
    return df


def format_millions(x: float, pos: int) -> str:
    return f"{x / 1e6:,.1f}M"


def plot_rolling_volume(df: pd.DataFrame, window: int = 7) -> Path:
    df = df.copy()
    df["rolling_volume"] = df["total_volume"].rolling(window=window, min_periods=1).sum()

    fig, ax = plt.subplots(figsize=(14, 9))
    ax.plot(df["date"], df["rolling_volume"], color="#0b3d91", linewidth=2)
    ax.scatter(df["date"], df["rolling_volume"], color="#0b3d91", s=18)

    ax.set_title(f"{window}-Day Rolling Total Trading Volume")
    ax.set_xlabel("Date")
    ax.set_ylabel(f"{window}-Day Total Volume (Contracts)")

    ax.yaxis.set_major_formatter(ticker.FuncFormatter(format_millions))
    ax.grid(color="0.85", linestyle="-", linewidth=0.8, alpha=0.8)

    fig.autofmt_xdate()
    fig.tight_layout()

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    output_path = OUTPUT_DIR / f"{window}d_rolling_total_volume.png"
    fig.savefig(output_path, dpi=150)
    plt.close(fig)
    return output_path


def main() -> None:
    files = sorted(DATA_DIR.glob("*.json"))
    if not files:
        raise SystemExit(f"No trade files found in {DATA_DIR}")

    df = collect_daily_volumes(files)

    rolling_plot_path = plot_rolling_volume(df, window=7)
    daily_csv_path = OUTPUT_DIR / "daily_total_volume_python.csv"
    df.to_csv(daily_csv_path, index=False, date_format="%Y-%m-%d")

    print("Daily totals saved to:", daily_csv_path)
    print("7-day rolling total chart saved to:", rolling_plot_path)


if __name__ == "__main__":
    main()

